File: swift/Tests/RezelTests/TreeTests.swift
import Testing
import Foundation
@testable import Rezel

private func makeTestTypes() -> (types: [NodeType], nodeSet: NodeSet, repeatType: NodeType) {
    let names = ["T", "a", "b", "c", "Pa", "Br"]
    var types: [NodeType] = []
    for (i, s) in names.enumerated() {
        let isAtom = s == "a" || s == "b" || s == "c"
        types.append(NodeType.define(spec: NodeType.DefineSpec(
            id: i,
            name: s,
            props: isAtom ? [nodePropGroup.add(match: ["atom": ["atom"]])] as [Any] : nil
        )))
    }
    let repeatType = NodeType.define(spec: NodeType.DefineSpec(id: types.count))
    types.append(repeatType)
    return (types, NodeSet(types: types), repeatType)
}

nonisolated(unsafe) private let testEnv = makeTestTypes()
nonisolated(unsafe) private let testTypes = testEnv.types
nonisolated(unsafe) private let testNodeSet = testEnv.nodeSet

private func id(_ n: String) -> Int {
    testTypes.first(where: { $0.name == n })!.id
}

private func mk(_ spec: String) -> Tree {
    var starts: [Int] = []
    var buffer: [Int] = []
    let chars = Array(spec)
    var pos = 0
    while pos < chars.count {
        let ch = chars[pos]
        if ch == "a" || ch == "b" || ch == "c" {
            let bufStart = buffer.count
            let groupStart = pos
            var i = 0
            while pos < chars.count {
                let c = chars[pos]
                guard c == "a" || c == "b" || c == "c" else { break }
                buffer.append(id(String(c)))
                buffer.append(pos)
                buffer.append(pos + 1)
                buffer.append(4)
                if i > 0 {
                    let curLen = buffer.count
                    buffer.append(testEnv.repeatType.id)
                    buffer.append(groupStart)
                    buffer.append(pos + 1)
                    buffer.append(curLen + 4 - bufStart)
                }
                i += 1
                pos += 1
            }
        } else if ch == "[" || ch == "(" {
            starts.append(buffer.count)
            starts.append(pos)
            pos += 1
        } else if ch == "]" || ch == ")" {
            let stringStart = starts.removeLast()
            let bufStartOff = starts.removeLast()
            let nodeName = ch == ")" ? "Pa" : "Br"
            let curLen = buffer.count
            buffer.append(id(nodeName))
            buffer.append(stringStart)
            buffer.append(pos + 1)
            buffer.append(curLen + 4 - bufStartOff)
            pos += 1
        } else {
            pos += 1
        }
    }
    return Tree.build(data: BuildData(
        buffer: buffer,
        nodeSet: testNodeSet,
        topID: 0,
        maxBufferLength: 10,
        minRepeatType: testEnv.repeatType.id
    ))
}

private func buildRecurSpec() -> String {
    func build(_ depth: Int) -> String {
        if depth > 0 {
            let inner = build(depth - 1)
            return "(" + inner + ")[" + inner + "]"
        } else {
            var result = ""
            for i in 0..<20 { result += ["a", "b", "c"][i % 3] }
            return result
        }
    }
    return build(6)
}

private let _recurLock = NSLock()
nonisolated(unsafe) private var _recur: Tree? = nil
private func recur() -> Tree {
    _recurLock.lock()
    defer { _recurLock.unlock() }
    if _recur == nil { _recur = mk(buildRecurSpec()) }
    return _recur!
}

private let _simpleLock = NSLock()
nonisolated(unsafe) private var _simple: Tree? = nil
private func simple() -> Tree {
    _simpleLock.lock()
    defer { _simpleLock.unlock() }
    if _simple == nil { _simple = mk("aaaa(bbb[ccc][aaa][()])") }
    return _simple!
}

nonisolated(unsafe) private let anonTree = Tree(
    type: testNodeSet.types[0],
    children: [
        Tree(
            type: NodeType.none,
            children: [
                Tree(type: testNodeSet.types[1], children: [], positions: [], length: 1),
                Tree(type: testNodeSet.types[2], children: [], positions: [], length: 1)
            ],
            positions: [0, 1],
            length: 2
        )
    ],
    positions: [0],
    length: 2
)

@Suite("SyntaxNode")
struct SyntaxNodeTests {
    @Test("can resolve at the top level")
    func resolveTopLevel() {
        var c = simple().resolve(pos: 2, side: -1)
        #expect(c.from == 1)
        #expect(c.to == 2)
        #expect(c.name == "a")
        #expect(c.parent!.name == "T")
        #expect(c.parent!.parent == nil)

        c = simple().resolve(pos: 2, side: 1)
        #expect(c.from == 2)
        #expect(c.to == 3)

        c = simple().resolve(pos: 2)
        #expect(c.name == "T")
        #expect(c.from == 0)
        #expect(c.to == 23)
    }

    @Test("can resolve deeper")
    func resolveDeeper() {
        let c = simple().resolve(pos: 10, side: 1)
        #expect(c.name == "c")
        #expect(c.from == 10)
        #expect(c.parent!.name == "Br")
        #expect(c.parent!.parent!.name == "Pa")
        #expect(c.parent!.parent!.parent!.name == "T")
    }

    @Test("can resolve in a large tree")
    func resolveLarge() {
        var c: SyntaxNode? = recur().resolve(pos: 10, side: 1)
        var depth = 1
        while c != nil { c = c!.parent; depth += 1 }
        #expect(depth == 9)
    }

    @Test("caches resolved parents")
    func cachesResolvedParents() {
        let a = recur().resolve(pos: 3, side: 1)
        let b = recur().resolve(pos: 3, side: 1)
        #expect(a as AnyObject === b as AnyObject)
    }

    @Test("skips anonymous nodes")
    func skipsAnonymous() {
        #expect(anonTree.description == "T(a,b)")
        #expect(anonTree.resolve(pos: 1).name == "T")
        #expect(anonTree.topNode.lastChild!.name == "b")
        #expect(anonTree.topNode.firstChild!.name == "a")
        #expect(anonTree.topNode.childAfter(1)!.name == "b")
    }

    @Test("allows access to the underlying tree")
    func accessTree() {
        let tree = mk("aaa[bbbbb(bb)bbbbbbb]aaa")
        var node = tree.topNode.firstChild!
        while node.name != "Br" { node = node.nextSibling! }
        #expect(node.tree != nil)
        #expect(node.tree!.type.name == "Br")
        node = node.firstChild!
        while node.name != "Pa" { node = node.nextSibling! }
        #expect(node.tree == nil)
        #expect(node.toTree().description == "Pa(b,b)")
        node = node.firstChild!
        #expect(node.name == "b")
        #expect(node.toTree().description == "b")
        #expect(node.toTree().children.count == 0)
    }
}

@Suite("TreeCursor")
struct TreeCursorTests {
    static let simpleCount: [String: Int] = ["a": 7, "b": 3, "c": 3, "Br": 3, "Pa": 2, "T": 1]

    @Test("iterates over all nodes")
    func iterateAll() {
        var count: [String: Int] = [:]
        var pos = 0
        let cur = simple().cursor()
        repeat {
            #expect(cur.from >= pos)
            pos = cur.from
            count[cur.name, default: 0] += 1
        } while cur.next()
        for (k, v) in Self.simpleCount {
            #expect(count[k] == v, "\(k): expected \(v), got \(count[k] ?? 0)")
        }
    }

    @Test("iterates over all nodes in reverse")
    func iterateReverse() {
        var count: [String: Int] = [:]
        var pos = 100
        let cur = simple().cursor()
        repeat {
            #expect(cur.to <= pos)
            pos = cur.to
            count[cur.name, default: 0] += 1
        } while cur.prev()
        for (k, v) in Self.simpleCount {
            #expect(count[k] == v, "\(k): expected \(v), got \(count[k] ?? 0)")
        }
    }

    @Test("works with internal iteration")
    func internalIteration() {
        var openCount: [String: Int] = [:]
        var closeCount: [String: Int] = [:]
        simple().iterate(enter: { t -> Bool in
            openCount[t.name, default: 0] += 1
            return true
        }, leave: { t in
            closeCount[t.name, default: 0] += 1
        })
        for (k, v) in Self.simpleCount {
            #expect(openCount[k] == v)
            #expect(closeCount[k] == v)
        }
    }

    @Test("handles iterating out of bounds")
    func outOfBounds() {
        var hit = 0
        Tree.empty.iterate(from: 0, to: 200, enter: { _ -> Bool in hit += 1; return true }, leave: { _ in hit += 1 })
        #expect(hit == 0)
    }

    @Test("internal iteration can be limited to a range")
    func limitedRange() {
        var seen: [String] = []
        simple().iterate(from: 3, to: 14, enter: { t -> Bool in
            seen.append(t.name)
            return t.name == "Br" ? false : true
        })
        #expect(seen.joined(separator: ",") == "T,a,a,Pa,b,b,b,Br,Br")
    }

    @Test("can leave nodes")
    func leaveNodes() {
        let c = simple().cursor()
        #expect(!c.parent())
        _ = c.next()
        _ = c.next()
        #expect(c.from == 1)
        #expect(c.parent())
        #expect(c.from == 0)
        for _ in 0..<6 { _ = c.next() }
        #expect(c.from == 5)
        #expect(c.parent())
        #expect(c.from == 4)
        #expect(c.parent())
        #expect(c.from == 0)
        #expect(!c.parent())
    }

    @Test("can move to a given position")
    func moveToPosition() {
        let tree = recur()
        let start = tree.length >> 1
        let cursor = tree.cursorAt(pos: start, side: 1)
        repeat {
            #expect(cursor.from >= start)
        } while cursor.next()
    }

    @Test("can move into a parent node")
    func moveIntoParent() {
        let c = simple().cursorAt(pos: 10)
        c.moveTo(pos: 2)
        #expect(c.name == "T")
    }

    @Test("isn't slow")
    func performance() {
        let tree = recur()
        let t0 = Date()
        var count = 0
        for _ in 0..<2000 {
            let c = tree.cursor()
            repeat {
                if c.from < 0 || c.name.isEmpty { Issue.record("BAD"); break }
                count += 1
            } while c.next()
        }
        let elapsed = Date().timeIntervalSince(t0) * 1000
        let perMS = Double(count) / max(elapsed, 1)
        #expect(perMS > 1000, "Performance too low: \(perMS) per ms")
    }

    @Test("can produce nodes")
    func produceNodes() {
        let node = simple().cursorAt(pos: 8, side: 1).node
        #expect(node.name == "Br")
        #expect(node.from == 8)
        #expect(node.parent!.name == "Pa")
        #expect(node.parent!.from == 4)
        #expect(node.parent!.parent!.name == "T")
        #expect(node.parent!.parent!.from == 0)
        #expect(node.parent!.parent!.parent == nil)
    }

    @Test("skips anonymous nodes in cursor")
    func skipsAnonymousCursor() {
        let c = anonTree.cursor()
        c.moveTo(pos: 1)
        #expect(c.name == "T")
        _ = c.firstChild()
        #expect(c.name == "a")
        _ = c.nextSibling()
        #expect(c.name == "b")
        #expect(!c.next())
    }
}

@Suite("matchContext")
struct MatchContextTests {
    @Test("can match on nodes")
    func matchNodes() {
        #expect(simple().resolve(pos: 10, side: 1).matchContext(["T", "Pa", "Br"]))
    }

    @Test("can match wildcards")
    func matchWildcards() {
        #expect(simple().resolve(pos: 10, side: 1).matchContext(["T", "", "Br"]))
    }

    @Test("can mismatch on nodes")
    func mismatchNodes() {
        #expect(!simple().resolve(pos: 10, side: 1).matchContext(["Q", "Br"]))
    }

    @Test("can match on cursor")
    func matchCursor() {
        let cur = simple().cursor()
        for _ in 0..<3 { _ = cur.enter(15, side: -1) }
        #expect(cur.matchContext(["T", "Pa", "Br"]))
    }
}

File: dotnet/Rezel.Tests/TreeTests.cs
using Rezel.Common;
using Microsoft.VisualStudio.TestTools.UnitTesting;

namespace Rezel.Tests;

[TestClass]
public class TreeTests
{
    static readonly NodeType[] Types;
    static readonly NodeSet NodeSet;
    static readonly NodeType RepeatType;

    static TreeTests()
    {
        var names = new[] { "T", "a", "b", "c", "Pa", "Br" };
        Types = new NodeType[names.Length + 1];
        for (var i = 0; i < names.Length; i++)
        {
            var isAtom = names[i] is "a" or "b" or "c";
            Types[i] = NodeType.Define(
                id: i,
                name: names[i],
                props: isAtom ? [NodeProps.Group.Add(_ => new[] { "atom" })] : null
            );
        }
        RepeatType = NodeType.Define(id: names.Length);
        Types[names.Length] = RepeatType;
        NodeSet = new NodeSet(Types);
    }

    static int Id(string n) => Types.First(x => x.Name == n).Id;

    static Tree Mk(string spec)
    {
        var starts = new List<int>();
        var buffer = new List<int>();
        var pos = 0;

        while (pos < spec.Length)
        {
            var ch = spec[pos];
            if (ch is 'a' or 'b' or 'c')
            {
                var bufStart = buffer.Count;
                var groupStart = pos;
                var i = 0;
                while (pos < spec.Length && spec[pos] is 'a' or 'b' or 'c')
                {
                    buffer.Add(Id(spec[pos].ToString()));
                    buffer.Add(pos);
                    buffer.Add(pos + 1);
                    buffer.Add(4);
                    if (i > 0)
                    {
                        var curLen = buffer.Count;
                        buffer.Add(RepeatType.Id);
                        buffer.Add(groupStart);
                        buffer.Add(pos + 1);
                        buffer.Add(curLen + 4 - bufStart);
                    }
                    i++;
                    pos++;
                }
            }
            else if (ch is '[' or '(')
            {
                starts.Add(buffer.Count);
                starts.Add(pos);
                pos++;
            }
            else if (ch is ']' or ')')
            {
                var stringStart = starts[^1]; starts.RemoveAt(starts.Count - 1);
                var bufStartOff = starts[^1]; starts.RemoveAt(starts.Count - 1);
                var nodeName = ch == ')' ? "Pa" : "Br";
                var curLen = buffer.Count;
                buffer.Add(Id(nodeName));
                buffer.Add(stringStart);
                buffer.Add(pos + 1);
                buffer.Add(curLen + 4 - bufStartOff);
                pos++;
            }
            else
            {
                pos++;
            }
        }

        return Tree.Build(new BuildData(
            buffer: buffer.ToArray(),
            nodeSet: NodeSet,
            topID: 0,
            maxBufferLength: 10,
            minRepeatType: RepeatType.Id
        ));
    }

    static string BuildRecurSpec(int depth)
    {
        if (depth > 0)
        {
            var inner = BuildRecurSpec(depth - 1);
            return "(" + inner + ")[" + inner + "]";
        }
        else
        {
            var result = "";
            for (var i = 0; i < 20; i++)
                result += new[] { "a", "b", "c" }[i % 3];
            return result;
        }
    }

    static Tree? _recur;
    static Tree Recur() => _recur ??= Mk(BuildRecurSpec(6));

    static Tree? _simple;
    static Tree Simple() => _simple ??= Mk("aaaa(bbb[ccc][aaa][()])");

    static Tree? _anonTree;
    static Tree AnonTree => _anonTree ??= new(
        Types[0],
        [
            new Tree(
                NodeType.None,
                [
                    new Tree(Types[1], [], [], 1),
                    new Tree(Types[2], [], [], 1),
                ],
                [0, 1],
                2
            ),
        ],
        [0],
        2
    );

    // SyntaxNode tests

    [TestMethod]
    public void CanResolveAtTopLevel()
    {
        var c = Simple().Resolve(2, -1);
        Assert.AreEqual(1, c.From);
        Assert.AreEqual(2, c.To);
        Assert.AreEqual("a", c.Name);
        Assert.AreEqual("T", c.Parent!.Name);
        Assert.IsNull(c.Parent!.Parent);

        c = Simple().Resolve(2, 1);
        Assert.AreEqual(2, c.From);
        Assert.AreEqual(3, c.To);

        c = Simple().Resolve(2);
        Assert.AreEqual("T", c.Name);
        Assert.AreEqual(0, c.From);
        Assert.AreEqual(23, c.To);
    }

    [TestMethod]
    public void CanResolveDeeper()
    {
        var c = Simple().Resolve(10, 1);
        Assert.AreEqual("c", c.Name);
        Assert.AreEqual(10, c.From);
        Assert.AreEqual("Br", c.Parent!.Name);
        Assert.AreEqual("Pa", c.Parent!.Parent!.Name);
        Assert.AreEqual("T", c.Parent!.Parent!.Parent!.Name);
    }

    [TestMethod]
    public void CanResolveInLargeTree()
    {
        SyntaxNode? c = Recur().Resolve(10, 1);
        var depth = 1;
        while (c != null)
        {
            c = c.Parent;
            depth++;
        }
        Assert.AreEqual(9, depth);
    }

    [TestMethod]
    public void CachesResolvedParents()
    {
        var a = Recur().Resolve(3, 1);
        var b = Recur().Resolve(3, 1);
        Assert.AreSame(a, b);
    }

    [TestMethod]
    public void SkipsAnonymousNodes()
    {
        Assert.AreEqual("T(a,b)", AnonTree.ToString());
        Assert.AreEqual("T", AnonTree.Resolve(1).Name);
        Assert.AreEqual("b", AnonTree.TopNode.LastChild!.Name);
        Assert.AreEqual("a", AnonTree.TopNode.FirstChild!.Name);
        Assert.AreEqual("b", AnonTree.TopNode.ChildAfter(1)!.Name);
    }

    [TestMethod]
    public void AllowsAccessToUnderlyingTree()
    {
        var tree = Mk("aaa[bbbbb(bb)bbbbbbb]aaa");
        var node = tree.TopNode.FirstChild!;
        while (node.Name != "Br") node = node.NextSibling!;
        Assert.IsNotNull(node.Tree);
        Assert.AreEqual("Br", node.Tree!.Type.Name);

        node = node.FirstChild!;
        while (node.Name != "Pa") node = node.NextSibling!;
        Assert.IsNull(node.Tree);
        Assert.AreEqual("Pa(b,b)", node.ToTree().ToString());

        node = node.FirstChild!;
        Assert.AreEqual("b", node.Name);
        Assert.AreEqual("b", node.ToTree().ToString());
        Assert.AreEqual(0, node.ToTree().Children.Length);
    }

    // TreeCursor tests

    static readonly Dictionary<string, int> SimpleCount = new()
    {
        ["a"] = 7,
        ["b"] = 3,
        ["c"] = 3,
        ["Br"] = 3,
        ["Pa"] = 2,
        ["T"] = 1
    };

    [TestMethod]
    public void CursorIteratesOverAllNodes()
    {
        var count = new Dictionary<string, int>();
        var pos = 0;
        var cur = Simple().Cursor();
        do
        {
            Assert.IsTrue(cur.From >= pos);
            pos = cur.From;
            count.TryGetValue(cur.Name, out var c);
            count[cur.Name] = c + 1;
        } while (cur.Next());
        foreach (var kv in SimpleCount)
            Assert.AreEqual(kv.Value, count.GetValueOrDefault(kv.Key, 0));
    }

    [TestMethod]
    public void CursorIteratesReverse()
    {
        var count = new Dictionary<string, int>();
        var pos = 100;
        var cur = Simple().Cursor();
        do
        {
            Assert.IsTrue(cur.To <= pos);
            pos = cur.To;
            count.TryGetValue(cur.Name, out var c);
            count[cur.Name] = c + 1;
        } while (cur.Prev());
        foreach (var kv in SimpleCount)
            Assert.AreEqual(kv.Value, count.GetValueOrDefault(kv.Key, 0));
    }

    [TestMethod]
    public void CursorWorksWithInternalIteration()
    {
        var openCount = new Dictionary<string, int>();
        var closeCount = new Dictionary<string, int>();
        Simple().Iterate(0, Simple().Length, IterMode.None,
            enter: t =>
            {
                openCount.TryGetValue(t.Name, out var c);
                openCount[t.Name] = c + 1;
                return true;
            },
            leave: t =>
            {
                closeCount.TryGetValue(t.Name, out var c);
                closeCount[t.Name] = c + 1;
            });
        foreach (var kv in SimpleCount)
        {
            Assert.AreEqual(kv.Value, openCount.GetValueOrDefault(kv.Key, 0));
            Assert.AreEqual(kv.Value, closeCount.GetValueOrDefault(kv.Key, 0));
        }
    }

    [TestMethod]
    public void CursorHandlesOutOfBounds()
    {
        var hit = 0;
        Tree.Empty.Iterate(0, 200, IterMode.None,
            enter: _ => { hit++; return true; },
            leave: _ => { hit++; });
        Assert.AreEqual(0, hit);
    }

    [TestMethod]
    public void CursorLimitedRange()
    {
        var seen = new List<string>();
        Simple().Iterate(3, 14, IterMode.None,
            enter: t =>
            {
                seen.Add(t.Name);
                return t.Name != "Br";
            });
        Assert.AreEqual("T,a,a,Pa,b,b,b,Br,Br", string.Join(",", seen));
    }

    [TestMethod]
    public void CursorCanLeaveNodes()
    {
        var c = Simple().Cursor();
        Assert.IsFalse(c.Parent());
        Assert.IsTrue(c.Next());
        Assert.IsTrue(c.Next());
        Assert.AreEqual(1, c.From);
        Assert.IsTrue(c.Parent());
        Assert.AreEqual(0, c.From);
        for (var j = 0; j < 6; j++) c.Next();
        Assert.AreEqual(5, c.From);
        Assert.IsTrue(c.Parent());
        Assert.AreEqual(4, c.From);
        Assert.IsTrue(c.Parent());
        Assert.AreEqual(0, c.From);
        Assert.IsFalse(c.Parent());
    }

    [TestMethod]
    public void CursorCanMoveToPosition()
    {
        var tree = Recur();
        var start = tree.Length >> 1;
        var cursor = tree.CursorAt(start, 1);
        do
        {
            Assert.IsTrue(cursor.From >= start);
        } while (cursor.Next());
    }

    [TestMethod]
    public void CursorCanMoveIntoParent()
    {
        var c = Simple().CursorAt(10);
        c.MoveTo(2);
        Assert.AreEqual("T", c.Name);
    }

    [TestMethod]
    public void CursorIsNotSlow()
    {
        var tree = Recur();
        var t0 = Environment.TickCount;
        var count = 0;
        for (var i = 0; i < 2000; i++)
        {
            var cur = tree.Cursor();
            do
            {
                if (cur.From < 0 || string.IsNullOrEmpty(cur.Name)) throw new Exception("BAD");
                count++;
            } while (cur.Next());
        }
        var elapsed = Math.Max(Environment.TickCount - t0, 1);
        var perMS = (double)count / elapsed;
        Assert.IsTrue(perMS > 10000, $"Performance too low: {perMS} per ms");
    }

    [TestMethod]
    public void CursorCanProduceNodes()
    {
        var node = Simple().CursorAt(8, 1).Node;
        Assert.AreEqual("Br", node.Name);
        Assert.AreEqual(8, node.From);
        Assert.AreEqual("Pa", node.Parent!.Name);
        Assert.AreEqual(4, node.Parent!.From);
        Assert.AreEqual("T", node.Parent!.Parent!.Name);
        Assert.AreEqual(0, node.Parent!.Parent!.From);
        Assert.IsNull(node.Parent!.Parent!.Parent);
    }

    [TestMethod]
    public void CursorSkipsAnonymousNodes()
    {
        var c = AnonTree.Cursor();
        c.MoveTo(1);
        Assert.AreEqual("T", c.Name);
        Assert.IsTrue(c.FirstChild());
        Assert.AreEqual("a", c.Name);
        Assert.IsTrue(c.NextSibling());
        Assert.AreEqual("b", c.Name);
        Assert.IsFalse(c.Next());
    }

    // matchContext tests

    [TestMethod]
    public void MatchContextCanMatchOnNodes()
    {
        Assert.IsTrue(Simple().Resolve(10, 1).MatchContext(["T", "Pa", "Br"]));
    }

    [TestMethod]
    public void MatchContextCanMatchWildcards()
    {
        Assert.IsTrue(Simple().Resolve(10, 1).MatchContext(["T", "", "Br"]));
    }

    [TestMethod]
    public void MatchContextCanMismatchOnNodes()
    {
        Assert.IsFalse(Simple().Resolve(10, 1).MatchContext(["Q", "Br"]));
    }

    [TestMethod]
    public void MatchContextCanMatchOnCursor()
    {
        var c = Simple().Cursor();
        for (var i = 0; i < 3; i++) c.Enter(15, -1);
        Assert.IsTrue(c.MatchContext(["T", "Pa", "Br"]));
    }
}
